django_mongodb_engine/base.py

```diff
@@ -234,49 +242,43 @@ def pop(name, default=None):
-                    exc_info = sys.exc_info()
-                    raise ImproperlyConfigured, exc_info[1], exc_info[2]
-
-                if user and password:
-                    if not self.database.authenticate(user, password):
-                        raise ImproperlyConfigured("Invalid username or password.")
-                          
-                self.connected = True
-                connection_created.send(sender=self.__class__, connection=self)
-                ''' execute a quick sample query so that the failure will happen 
-                    on the command that is run after the switchover, auth succeeds
-                    on secondary but commands cannot be run. This command will
-                    throw an exception and hence we will attempt to reconnect again '''
-                self.database['system.indexes'].find_one()
-                break
-            except Exception as e:
-                print 'MongoConnectionFailure to database %s %s' % (db_name,str(e))
-                print traceback.format_exc()
-
-                ''' Make sure we set connected to False just in case we failed on the send '''
-                self.connected = False
-                
-                ''' initialize these instance variables, so that we can delete them '''
-                ''' this will ensure that the __get_attr__ class method properly reconnects '''
-                self.database = None
-                self.connection = None
-                del self.database
-                del self.connection
-
-                attempts += 1
-                if attempts < self.conn_retries:
-                    time.sleep(self.conn_sleep_interval)
-                    print 'MongoConnectionRetry attempt=%d' % attempts
-                    continue
-                raise e
-            
+        read_preference = options.get('read_preference')
+
+        if not read_preference:
+            read_preference = options.get('slave_okay', options.get('slaveok'))
+            if read_preference:
+                options['read_preference'] = ReadPreference.SECONDARY
+                warnings.warn("slave_okay has been deprecated. "
+                              "Please use read_preference instead.")
+
+        conn_options = dict(
+            host=host,
+            port=int(port),
+            document_class=dict,
+            tz_aware=False,
+            connect=False
+        )
+        conn_options.update(options)
+
+        if user and password:
+            conn_options.update(dict(
+                username=user,
+                password=password,
+                authSource=db_name
+            ))
+
+        try:
+            self.connection = MongoClient(**conn_options)
+            self.database = self.connection[db_name]
+            # Make a simple query so that connection can be established.
+            # Note: Making a call that requires authentication.
+            self.database['system.indexes'].find_one()
```

Comment:
Changes for PyMongo3.6.0 from L#263 to here. Note: Our changes are not backwards compatible.